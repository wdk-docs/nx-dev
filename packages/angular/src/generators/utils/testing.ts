import type { Tree } from '@nrwl/devkit';
import { wrapAngularDevkitSchematic } from '@nrwl/devkit/ngcli-adapter';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { Linter } from '@nrwl/linter';
import { UnitTestRunner } from '../../utils/test-runners';
import libraryGenerator from '../library/library';

export async function createStorybookTestWorkspaceForLib(
  libName: string
): Promise<Tree> {
  let tree = createTreeWithEmptyWorkspace();

  const moduleGenerator = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'module'
  );
  const componentGenerator = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'component'
  );

  await libraryGenerator(tree, {
    name: libName,
    buildable: false,
    linter: Linter.EsLint,
    publishable: false,
    simpleModuleName: false,
    skipFormat: false,
    unitTestRunner: UnitTestRunner.Jest,
  });

  await componentGenerator(tree, {
    name: 'test-button',
    project: libName,
  });

  tree.write(
    `libs/${libName}/src/lib/test-button/test-button.component.ts`,
    `import { Component, Input } from '@angular/core';

export type ButtonStyle = 'default' | 'primary' | 'accent';

@Component({
  selector: 'proj-test-button',
  templateUrl: './test-button.component.html',
  styleUrls: ['./test-button.component.css']
})
export class TestButtonComponent {
  @Input('buttonType') type = 'button';
  @Input() style: ButtonStyle = 'default';
  @Input() age?: number;
  @Input() isOn = false;
}`
  );

  tree.write(
    `libs/${libName}/src/lib/test-button/test-button.component.html`,
    `<button [attr.type]="type" [ngClass]="style"></button>`
  );

  const modulePath = `libs/${libName}/src/lib/${libName}.module.ts`;
  tree.write(
    modulePath,
    `import * as ButtonExports from './test-button/test-button.component';
    ${tree.read(modulePath)}`
  );

  // create a module with component that gets exported in a barrel file
  await moduleGenerator(tree, {
    name: 'barrel',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'barrel-button',
    project: libName,
    path: `libs/${libName}/src/lib/barrel`,
    module: 'barrel',
  });

  tree.write(
    `libs/${libName}/src/lib/barrel/barrel-button/index.ts`,
    `export * from './barrel-button.component';`
  );

  tree.write(
    `libs/${libName}/src/lib/barrel/barrel.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BarrelButtonComponent } from './barrel-button';

@NgModule({
  imports: [CommonModule],
  declarations: [BarrelButtonComponent],
})
export class BarrelModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  await moduleGenerator(tree, {
    name: 'variable-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-declare-button',
    project: libName,
    path: `libs/${libName}/src/lib/variable-declare`,
    module: 'variable-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-declare-view',
    project: libName,
    path: `libs/${libName}/src/lib/variable-declare`,
    module: 'variable-declare',
  });

  tree.write(
    `libs/${libName}/src/lib/variable-declare/variable-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableDeclareButtonComponent } from './variable-declare-button/variable-declare-button.component';
import { VariableDeclareViewComponent } from './variable-declare-view/variable-declare-view.component';

const COMPONENTS = [
  VariableDeclareButtonComponent,
  VariableDeclareViewComponent
]

@NgModule({
  imports: [CommonModule],
  declarations: COMPONENTS,
  exports: COMPONENTS
})
export class VariableDeclareModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  await moduleGenerator(tree, {
    name: 'variable-spread-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-button',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-view',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-anotherview',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  tree.write(
    `libs/${libName}/src/lib/variable-spread-declare/variable-spread-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableSpreadDeclareButtonComponent } from './variable-spread-declare-button/variable-spread-declare-button.component';
import { VariableSpreadDeclareViewComponent } from './variable-spread-declare-view/variable-spread-declare-view.component';
import { VariableSpreadDeclareAnotherviewComponent } from './variable-spread-declare-anotherview/variable-spread-declare-anotherview.component';

const COMPONENTS = [ 
  VariableSpreadDeclareButtonComponent, 
  VariableSpreadDeclareViewComponent 
]

@NgModule({
  imports: [CommonModule],
  declarations: [...COMPONENTS, VariableSpreadDeclareAnotherviewComponent],
})
export class VariableSpreadDeclareModule {}`
  );

  // create a module where declared components are pulled from a static member of the module
  await moduleGenerator(tree, {
    name: 'static-member-declarations',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'cmp1',
    project: libName,
    path: `libs/${libName}/src/lib/static-member-declarations`,
    module: 'static-member-declarations',
  });

  await componentGenerator(tree, {
    name: 'cmp2',
    project: libName,
    path: `libs/${libName}/src/lib/static-member-declarations`,
    module: 'static-member-declarations',
  });

  tree.write(
    `libs/${libName}/src/lib/static-member-declarations/static-member-declarations.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Cmp1Component } from './cmp1/cmp1.component';
import { Cmp2Component } from './cmp2/cmp2.component';

@NgModule({
  imports: [CommonModule],
  declarations: StaticMemberDeclarationsModule.COMPONENTS,
  exports: StaticMemberDeclarationsModule.COMPONENTS
})
export class StaticMemberDeclarationsModule {
  static readonly COMPONENTS = [Cmp1Component, Cmp2Component];
}`
  );

  // create another button in a nested subpath
  await moduleGenerator(tree, {
    name: 'nested',
    project: libName,
    path: `libs/${libName}/src/lib`,
  });

  await componentGenerator(tree, {
    name: 'nested-button',
    project: libName,
    module: 'nested',
    path: `libs/${libName}/src/lib/nested`,
  });

  await componentGenerator(tree, {
    name: 'test-other',
    project: libName,
  });

  return tree;
}
