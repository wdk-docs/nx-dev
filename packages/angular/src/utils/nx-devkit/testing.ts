import type { Tree } from '@nrwl/devkit';
import { names, updateJson } from '@nrwl/devkit';

export interface AppConfig {
  appName: string; // name of app
  appModule: string; // app/app.module.ts in the above sourceDir
}
export interface LibConfig {
  name: string;
  module: string;
  barrel: string;
}

var appConfig: AppConfig; // configure built in createApp()
var libConfig: LibConfig;

export function getAppConfig(): AppConfig {
  return appConfig;
}
export function getLibConfig(): LibConfig {
  return libConfig;
}

export function createApp(
  tree: Tree,
  appName: string,
  routing: boolean = true
) {
  appName = names(appName).fileName;
  // save for getAppDir() lookup by external *.spec.ts tests
  appConfig = {
    appName,
    appModule: `/apps/${appName}/src/app/app.module.ts`,
  };

  tree.write(
    appConfig.appModule,
    `
     import { NgModule } from '@angular/core';
     import { BrowserModule } from '@angular/platform-browser';
     ${routing ? "import { RouterModule } from '@angular/router'" : ''};
     import { AppComponent } from './app.component';
     @NgModule({
       imports: [BrowserModule, ${routing ? 'RouterModule.forRoot([])' : ''}],
       declarations: [AppComponent],
       bootstrap: [AppComponent]
     })
     export class AppModule {}
  `
  );
  tree.write(
    `/apps/${appName}/src/main.ts`,
    `
    import { enableProdMode } from '@angular/core';
    import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';

    import { AppModule } from './app/app.module';
    import { environment } from './environments/environment';

    if (environment.production) {
      enableProdMode();
    }

    platformBrowserDynamic()
      .bootstrapModule(AppModule)
      .catch(err => console.log(err));
  `
  );
  tree.write(
    `/apps/${appName}/tsconfig.app.json`,
    JSON.stringify({
      include: ['**/*.ts'],
    })
  );
  tree.write(
    `/apps/${appName}-e2e/tsconfig.e2e.json`,
    JSON.stringify({
      include: ['../**/*.ts'],
    })
  );
  tree.write(
    '/workspace.json',
    JSON.stringify({
      newProjectRoot: '',
      version: 1,
      projects: {
        [appName]: {
          root: `apps/${appName}`,
          sourceRoot: `apps/${appName}/src`,
          architect: {
            build: {
              options: {
                main: `apps/${appName}/src/main.ts`,
              },
            },
            serve: {
              options: {},
            },
          },
          tags: [],
        },
      },
    })
  );
}

export function createLib(tree: Tree, libName: string) {
  const { name, className, fileName, propertyName } = names(libName);

  libConfig = {
    name,
    module: `/libs/${propertyName}/src/lib/${fileName}.module.ts`,
    barrel: `/libs/${propertyName}/src/index.ts`,
  };

  tree.write(
    libConfig.module,
    `
      import { NgModule } from '@angular/core';
      import { CommonModule } from '@angular/common';
      @NgModule({
        imports: [
          CommonModule
        ],
        providers: []
      })
      export class ${className}Module { }
  `
  );
  tree.write(
    libConfig.barrel,
    `
    export * from './lib/${fileName}.module';
  `
  );
}
