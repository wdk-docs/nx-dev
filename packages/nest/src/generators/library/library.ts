import type { GeneratorCallback, Tree } from '@nrwl/devkit';
import { convertNxGenerator, formatFiles } from '@nrwl/devkit';
import { libraryGenerator as nodeLibraryGenerator } from '@nrwl/node';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import { addDependencies } from '../init/lib';
import {
  addExportsToBarrelFile,
  addProject,
  createFiles,
  deleteFiles,
  normalizeOptions,
  toNodeLibraryGeneratorOptions,
  updateTsConfig,
} from './lib';
import type { LibraryGeneratorOptions } from './schema';

export async function libraryGenerator(
  tree: Tree,
  rawOptions: LibraryGeneratorOptions
): Promise<GeneratorCallback> {
  const options = normalizeOptions(tree, rawOptions);
  const nodeLibraryTask = await nodeLibraryGenerator(
    tree,
    toNodeLibraryGeneratorOptions(options)
  );
  const installDepsTask = addDependencies(tree);
  deleteFiles(tree, options);
  createFiles(tree, options);
  addExportsToBarrelFile(tree, options);
  updateTsConfig(tree, options);
  addProject(tree, options);

  if (!options.skipFormat) {
    await formatFiles(tree);
  }

  return runTasksInSerial(nodeLibraryTask, installDepsTask);
}

export default libraryGenerator;

export const librarySchematic = convertNxGenerator(libraryGenerator);
