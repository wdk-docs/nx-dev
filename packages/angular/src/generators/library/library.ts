import {
  addDependenciesToPackageJson,
  formatFiles,
  installPackagesTask,
  moveFilesToNewDirectory,
  Tree,
} from '@nrwl/devkit';
import { wrapAngularDevkitSchematic } from '@nrwl/devkit/ngcli-adapter';
import { jestProjectGenerator } from '@nrwl/jest';
import { Linter } from '@nrwl/linter';
import { convertToNxProjectGenerator } from '@nrwl/workspace';
import init from '../../generators/init/init';
import { postcssVersion } from '../../utils/versions';
import addLintingGenerator from '../add-linting/add-linting';
import karmaProjectGenerator from '../karma-project/karma-project';
import setupTailwindGenerator from '../setup-tailwind/setup-tailwind';
import { addModule } from './lib/add-module';
import {
  enableStrictTypeChecking,
  setLibraryStrictDefault,
} from './lib/enable-strict-type-checking';
import { normalizeOptions } from './lib/normalize-options';
import { NormalizedSchema } from './lib/normalized-schema';
import { updateLibPackageNpmScope } from './lib/update-lib-package-npm-scope';
import { updateProject } from './lib/update-project';
import { updateTsConfig } from './lib/update-tsconfig';
import { Schema } from './schema';

export async function libraryGenerator(host: Tree, schema: Partial<Schema>) {
  // Do some validation checks
  if (!schema.routing && schema.lazy) {
    throw new Error(`To use "--lazy" option, "--routing" must also be set.`);
  }

  if (schema.publishable === true && !schema.importPath) {
    throw new Error(
      `For publishable libs you have to provide a proper "--importPath" which needs to be a valid npm package name (e.g. my-awesome-lib or @myorg/my-lib)`
    );
  }

  if (schema.addTailwind && !schema.buildable && !schema.publishable) {
    throw new Error(
      `To use "--addTailwind" option, you have to set either "--buildable" or "--publishable".`
    );
  }

  const options = normalizeOptions(host, schema);

  await init(host, {
    ...options,
    skipFormat: true,
  });

  const runAngularLibrarySchematic = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'library'
  );
  await runAngularLibrarySchematic(host, {
    name: options.name,
    prefix: options.prefix,
    entryFile: 'index',
    skipPackageJson: !(options.publishable || options.buildable),
    skipTsConfig: true,
  });

  moveFilesToNewDirectory(host, options.name, options.projectRoot);
  await updateProject(host, options);
  updateTsConfig(host, options);
  await addUnitTestRunner(host, options);
  updateNpmScopeIfBuildableOrPublishable(host, options);
  addModule(host, options);
  setStrictMode(host, options);
  await addLinting(host, options);

  if (options.addTailwind) {
    await setupTailwindGenerator(host, {
      project: options.name,
      skipFormat: true,
    });
  }

  if (options.buildable || options.publishable) {
    addDependenciesToPackageJson(
      host,
      {},
      {
        postcss: postcssVersion,
        'postcss-import': '^14.0.2',
        'postcss-preset-env': '^6.7.0',
        'postcss-url': '^10.1.1',
      }
    );
  }

  if (options.standaloneConfig) {
    await convertToNxProjectGenerator(host, {
      project: options.name,
      all: false,
      skipFormat: true,
    });
  }

  if (!options.skipFormat) {
    await formatFiles(host);
  }

  return () => {
    installPackagesTask(host);
  };
}

async function addUnitTestRunner(host: Tree, options: NormalizedSchema) {
  if (options.unitTestRunner === 'jest') {
    await jestProjectGenerator(host, {
      project: options.name,
      setupFile: 'angular',
      supportTsx: false,
      skipSerializers: false,
      skipFormat: true,
    });
  } else if (options.unitTestRunner === 'karma') {
    await karmaProjectGenerator(host, {
      project: options.name,
      skipFormat: true,
    });
  }
}

function updateNpmScopeIfBuildableOrPublishable(
  host: Tree,
  options: NormalizedSchema
) {
  if (options.buildable || options.publishable) {
    updateLibPackageNpmScope(host, options);
  }
}

function setStrictMode(host: Tree, options: NormalizedSchema) {
  if (options.strict) {
    enableStrictTypeChecking(host, options);
  } else {
    setLibraryStrictDefault(host, options.strict);
  }
}

async function addLinting(host: Tree, options: NormalizedSchema) {
  if (options.linter === Linter.None) {
    return;
  }
  await addLintingGenerator(host, {
    projectName: options.name,
    projectRoot: options.projectRoot,
    prefix: options.prefix,
    setParserOptionsProject: options.setParserOptionsProject,
    skipFormat: true,
  });
}

export default libraryGenerator;
