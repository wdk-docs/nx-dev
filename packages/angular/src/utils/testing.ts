import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Rule, Tree } from '@angular-devkit/schematics';

const testRunner = new SchematicTestRunner(
  '@nrwl/angular',
  join(__dirname, '../../generators.json')
);

testRunner.registerCollection(
  '@nrwl/jest',
  join(__dirname, '../../../jest/generators.json')
);

testRunner.registerCollection(
  '@nrwl/workspace',
  join(__dirname, '../../../workspace/generators.json')
);

testRunner.registerCollection(
  '@nrwl/cypress',
  join(__dirname, '../../../cypress/generators.json')
);

testRunner.registerCollection(
  '@nrwl/storybook',
  join(__dirname, '../../../storybook/generators.json')
);

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/workspace',
  join(__dirname, '../../migrations.json')
);

export function runMigration<SchemaOptions = any>(
  schematicName: string,
  options: SchemaOptions,
  tree: Tree
) {
  return migrationTestRunner
    .runSchematicAsync(schematicName, options, tree)
    .toPromise();
}

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}
