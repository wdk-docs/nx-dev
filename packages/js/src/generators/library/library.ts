import { convertNxGenerator, getWorkspaceLayout, Tree } from '@nrwl/devkit';
import { join } from 'path';
import { projectGenerator } from '../../utils/project-generator';
import { GeneratorSchema } from '../../utils/schema';

export async function libraryGenerator(tree: Tree, schema: GeneratorSchema) {
  const { libsDir } = getWorkspaceLayout(tree);
  return projectGenerator(
    tree,
    schema,
    libsDir,
    join(__dirname, './files'),
    'library'
  );
}

export default libraryGenerator;
export const librarySchematic = convertNxGenerator(libraryGenerator);
