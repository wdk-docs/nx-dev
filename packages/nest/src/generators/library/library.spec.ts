import type { NxJsonConfiguration, Tree } from '@nrwl/devkit';
import * as devkit from '@nrwl/devkit';
import { readJson, readProjectConfiguration } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { libraryGenerator } from './library';

describe('lib', () => {
  let tree: Tree;
  const libFileName = 'my-lib';
  const libName = 'myLib';

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    jest.clearAllMocks();
  });

  describe('not nested', () => {
    it('should update workspace.json', async () => {
      await libraryGenerator(tree, { name: libName });

      const workspaceJson = readJson(tree, '/workspace.json');
      expect(workspaceJson.projects[libFileName].root).toEqual(
        `libs/${libFileName}`
      );
      expect(
        workspaceJson.projects[libFileName].architect.build
      ).toBeUndefined();
      expect(workspaceJson.projects[libFileName].architect.lint).toEqual({
        builder: '@nrwl/linter:eslint',
        outputs: ['{options.outputFile}'],
        options: {
          lintFilePatterns: [`libs/${libFileName}/**/*.ts`],
        },
      });
      expect(workspaceJson.projects[libFileName].architect.test).toEqual({
        builder: '@nrwl/jest:jest',
        outputs: [`coverage/libs/${libFileName}`],
        options: {
          jestConfig: `libs/${libFileName}/jest.config.js`,
          passWithNoTests: true,
        },
      });
    });

    it('should include a controller', async () => {
      await libraryGenerator(tree, { name: libName, controller: true });

      expect(
        tree.exists(`libs/${libFileName}/src/lib/${libFileName}.controller.ts`)
      ).toBeTruthy();
    });

    it('should include a service', async () => {
      await libraryGenerator(tree, { name: libName, service: true });

      expect(
        tree.exists(`libs/${libFileName}/src/lib/${libFileName}.service.ts`)
      ).toBeTruthy();
    });

    it('should add the @Global decorator', async () => {
      await libraryGenerator(tree, { name: libName, global: true });

      expect(
        tree.read(
          `libs/${libFileName}/src/lib/${libFileName}.module.ts`,
          'utf-8'
        )
      ).toMatchSnapshot();
    });

    it('should remove the default file from @nrwl/node:lib', async () => {
      await libraryGenerator(tree, { name: libName, global: true });

      expect(
        tree.exists(`libs/${libFileName}/src/lib/${libFileName}.spec.ts`)
      ).toBeFalsy();
      expect(
        tree.exists(`libs/${libFileName}/src/lib/${libFileName}.ts`)
      ).toBeFalsy();
    });

    it('should provide the controller and service', async () => {
      await libraryGenerator(tree, {
        name: libName,
        controller: true,
        service: true,
      });

      console.log(
        tree.read(
          `libs/${libFileName}/src/lib/${libFileName}.controller.spec.ts`,
          'utf-8'
        )
      );

      expect(
        tree.read(
          `libs/${libFileName}/src/lib/${libFileName}.module.ts`,
          'utf-8'
        )
      ).toMatchSnapshot();
      expect(
        tree.read(
          `libs/${libFileName}/src/lib/${libFileName}.controller.ts`,
          'utf-8'
        )
      ).toMatchSnapshot();
      expect(
        tree.read(`libs/${libFileName}/src/index.ts`, 'utf-8')
      ).toMatchSnapshot();
    });

    it('should update tags', async () => {
      await libraryGenerator(tree, { name: libName, tags: 'one,two' });

      const projects = Object.fromEntries(devkit.getProjects(tree));
      expect(projects).toEqual({
        [libFileName]: expect.objectContaining({
          tags: ['one', 'two'],
        }),
      });
    });

    it('should update root tsconfig.json', async () => {
      await libraryGenerator(tree, { name: libName });

      const tsconfigJson = readJson(tree, '/tsconfig.base.json');
      expect(
        tsconfigJson.compilerOptions.paths[`@proj/${libFileName}`]
      ).toEqual([`libs/${libFileName}/src/index.ts`]);
    });

    it('should create a local tsconfig.json', async () => {
      await libraryGenerator(tree, { name: libName });

      const tsconfigJson = readJson(tree, `libs/${libFileName}/tsconfig.json`);
      expect(tsconfigJson).toMatchSnapshot();
    });

    it('should extend the local tsconfig.json with tsconfig.spec.json', async () => {
      await libraryGenerator(tree, { name: libName });

      const tsconfigJson = readJson(
        tree,
        `libs/${libFileName}/tsconfig.spec.json`
      );
      expect(tsconfigJson.extends).toEqual('./tsconfig.json');
    });

    it('should extend the local tsconfig.json with tsconfig.lib.json', async () => {
      await libraryGenerator(tree, { name: libName });

      const tsconfigJson = readJson(
        tree,
        `libs/${libFileName}/tsconfig.lib.json`
      );
      expect(tsconfigJson.extends).toEqual('./tsconfig.json');
      expect(tsconfigJson.exclude).toEqual(['**/*.spec.ts', '**/*.test.ts']);
    });

    it('should generate files', async () => {
      await libraryGenerator(tree, { name: libName });

      expect(tree.exists(`libs/${libFileName}/jest.config.js`)).toBeTruthy();
      expect(tree.exists(`libs/${libFileName}/src/index.ts`)).toBeTruthy();
      expect(
        tree.exists(`libs/${libFileName}/src/lib/${libFileName}.spec.ts`)
      ).toBeFalsy();
      expect(
        readJson(tree, `libs/${libFileName}/.eslintrc.json`)
      ).toMatchSnapshot();
    });
  });

  describe('nested', () => {
    const dirName = 'myDir';
    const dirFileName = 'my-dir';
    const nestedLibFileName = `${dirFileName}-${libFileName}`;

    it('should update tags', async () => {
      await libraryGenerator(tree, {
        name: libName,
        directory: dirName,
        tags: 'one,two',
      });

      const projects = Object.fromEntries(devkit.getProjects(tree));
      expect(projects).toEqual({
        [nestedLibFileName]: expect.objectContaining({ tags: ['one', 'two'] }),
      });
    });

    it('should generate files', async () => {
      await libraryGenerator(tree, { name: libName, directory: dirName });

      expect(
        tree.exists(`libs/${dirFileName}/${libFileName}/jest.config.js`)
      ).toBeTruthy();
      expect(
        tree.exists(`libs/${dirFileName}/${libFileName}/src/index.ts`)
      ).toBeTruthy();
      expect(
        tree.exists(
          `libs/${dirFileName}/${libFileName}/src/lib/${libFileName}.spec.ts`
        )
      ).toBeFalsy();
    });

    it('should update workspace.json', async () => {
      await libraryGenerator(tree, { name: libName, directory: dirName });

      const project = readProjectConfiguration(tree, nestedLibFileName);
      expect(project.root).toEqual(`libs/${dirFileName}/${libFileName}`);
      expect(project.targets.lint).toEqual({
        executor: '@nrwl/linter:eslint',
        outputs: ['{options.outputFile}'],
        options: {
          lintFilePatterns: [`libs/${dirFileName}/${libFileName}/**/*.ts`],
        },
      });
    });

    it('should update tsconfig.json', async () => {
      await libraryGenerator(tree, { name: libName, directory: dirName });

      const tsconfigJson = readJson(tree, '/tsconfig.base.json');
      expect(
        tsconfigJson.compilerOptions.paths[
          `@proj/${dirFileName}/${libFileName}`
        ]
      ).toEqual([`libs/${dirFileName}/${libFileName}/src/index.ts`]);
      expect(
        tsconfigJson.compilerOptions.paths[`${nestedLibFileName}/*`]
      ).toBeUndefined();
    });

    it('should create a local tsconfig.json', async () => {
      await libraryGenerator(tree, { name: libName, directory: dirName });

      expect(
        readJson(tree, `libs/${dirFileName}/${libFileName}/tsconfig.json`)
      ).toMatchSnapshot();
    });
  });

  describe('--strict', () => {
    it('should update the projects tsconfig with strict true', async () => {
      await libraryGenerator(tree, { name: libName, strict: true });

      const tsconfigJson = readJson(
        tree,
        `/libs/${libFileName}/tsconfig.lib.json`
      );
      expect(tsconfigJson.compilerOptions.strict).toBe(true);
      expect(
        tsconfigJson.compilerOptions.forceConsistentCasingInFileNames
      ).toBe(true);
      expect(tsconfigJson.compilerOptions.noImplicitReturns).toBe(true);
      expect(tsconfigJson.compilerOptions.noFallthroughCasesInSwitch).toBe(
        true
      );
    });

    it('should default to strict false', async () => {
      await libraryGenerator(tree, { name: libName });

      const tsconfigJson = readJson(
        tree,
        `/libs/${libFileName}/tsconfig.lib.json`
      );
      expect(tsconfigJson.compilerOptions.strict).not.toBeDefined();
      expect(
        tsconfigJson.compilerOptions.forceConsistentCasingInFileNames
      ).not.toBeDefined();
      expect(tsconfigJson.compilerOptions.noImplicitReturns).not.toBeDefined();
      expect(
        tsconfigJson.compilerOptions.noFallthroughCasesInSwitch
      ).not.toBeDefined();
    });
  });

  describe('--unit-test-runner none', () => {
    it('should not generate test configuration', async () => {
      await libraryGenerator(tree, { name: libName, unitTestRunner: 'none' });

      expect(tree.exists(`libs/${libFileName}/tsconfig.spec.json`)).toBeFalsy();
      expect(tree.exists(`libs/${libFileName}/jest.config.js`)).toBeFalsy();
      expect(
        tree.exists(`libs/${libFileName}/lib/${libFileName}.spec.ts`)
      ).toBeFalsy();
      expect(
        readJson(tree, `libs/${libFileName}/tsconfig.json`)
      ).toMatchSnapshot();
      const project = readProjectConfiguration(tree, libFileName);
      expect(project.targets.test).toBeUndefined();
      expect(project.targets.lint).toMatchSnapshot();
    });
  });

  describe('publishable package', () => {
    it('should update package.json', async () => {
      const importPath = `@proj/${libName}`;

      await libraryGenerator(tree, {
        name: libName,
        publishable: true,
        importPath,
      });

      const packageJson = readJson(tree, `libs/${libFileName}/package.json`);
      expect(packageJson.name).toEqual(importPath);
    });
  });

  describe('compiler options target', () => {
    it('should set target to es6 in tsconfig.lib.json by default', async () => {
      await libraryGenerator(tree, { name: libName });

      const tsconfigJson = readJson(
        tree,
        `libs/${libFileName}/tsconfig.lib.json`
      );
      expect(tsconfigJson.compilerOptions.target).toEqual('es6');
    });

    it('should set target to es2020 in tsconfig.lib.json', async () => {
      await libraryGenerator(tree, { name: libName, target: 'es2020' });

      const tsconfigJson = readJson(
        tree,
        `libs/${libFileName}/tsconfig.lib.json`
      );
      expect(tsconfigJson.compilerOptions.target).toEqual('es2020');
    });
  });

  describe('--skipFormat', () => {
    it('should format files by default', async () => {
      jest.spyOn(devkit, 'formatFiles');

      await libraryGenerator(tree, { name: libName });

      expect(devkit.formatFiles).toHaveBeenCalled();
    });

    it('should not format files when --skipFormat=true', async () => {
      jest.spyOn(devkit, 'formatFiles');

      await libraryGenerator(tree, { name: libName, skipFormat: true });

      expect(devkit.formatFiles).not.toHaveBeenCalled();
    });
  });

  describe('--testEnvironment', () => {
    it('should set target jest testEnvironment to node by default', async () => {
      await libraryGenerator(tree, { name: libName });

      expect(
        tree.read(`libs/${libFileName}/jest.config.js`, 'utf-8')
      ).toMatchSnapshot();
    });

    it('should set target jest testEnvironment to jsdom', async () => {
      await libraryGenerator(tree, { name: libName, testEnvironment: 'jsdom' });

      expect(
        tree.read(`libs/${libFileName}/jest.config.js`, 'utf-8')
      ).toMatchSnapshot();
    });
  });
});
