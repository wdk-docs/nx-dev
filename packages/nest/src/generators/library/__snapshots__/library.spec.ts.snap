// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib --testEnvironment should set target jest testEnvironment to jsdom 1`] = `
"module.exports = {
  displayName: 'my-lib',
  preset: '../../jest.preset.js',
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    }
  },
  testEnvironment: 'node',
  transform: {
    '^.+\\\\\\\\.[tj]sx?$': 'ts-jest'
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/my-lib'
};
"
`;

exports[`lib --testEnvironment should set target jest testEnvironment to node by default 1`] = `
"module.exports = {
  displayName: 'my-lib',
  preset: '../../jest.preset.js',
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    }
  },
  testEnvironment: 'node',
  transform: {
    '^.+\\\\\\\\.[tj]sx?$': 'ts-jest'
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/my-lib'
};
"
`;

exports[`lib --unit-test-runner none should not generate test configuration 1`] = `
Object {
  "extends": "../../tsconfig.base.json",
  "files": Array [],
  "include": Array [],
  "references": Array [
    Object {
      "path": "./tsconfig.lib.json",
    },
  ],
}
`;

exports[`lib --unit-test-runner none should not generate test configuration 2`] = `
Object {
  "executor": "@nrwl/linter:eslint",
  "options": Object {
    "lintFilePatterns": Array [
      "libs/my-lib/**/*.ts",
    ],
  },
  "outputs": Array [
    "{options.outputFile}",
  ],
}
`;

exports[`lib nested should create a local tsconfig.json 1`] = `
Object {
  "extends": "../../../tsconfig.base.json",
  "files": Array [],
  "include": Array [],
  "references": Array [
    Object {
      "path": "./tsconfig.lib.json",
    },
    Object {
      "path": "./tsconfig.spec.json",
    },
  ],
}
`;

exports[`lib not nested should add the @Global decorator 1`] = `
"import { Module, Global } from '@nestjs/common';

@Global()
@Module({
	controllers: [],
	providers: [],
	exports: [],
})
export class MyLibModule {}
"
`;

exports[`lib not nested should create a local tsconfig.json 1`] = `
Object {
  "extends": "../../tsconfig.base.json",
  "files": Array [],
  "include": Array [],
  "references": Array [
    Object {
      "path": "./tsconfig.lib.json",
    },
    Object {
      "path": "./tsconfig.spec.json",
    },
  ],
}
`;

exports[`lib not nested should generate files 1`] = `
Object {
  "extends": Array [
    "../../.eslintrc.json",
  ],
  "ignorePatterns": Array [
    "!**/*",
  ],
  "overrides": Array [
    Object {
      "files": Array [
        "*.ts",
        "*.tsx",
        "*.js",
        "*.jsx",
      ],
      "rules": Object {},
    },
    Object {
      "files": Array [
        "*.ts",
        "*.tsx",
      ],
      "rules": Object {},
    },
    Object {
      "files": Array [
        "*.js",
        "*.jsx",
      ],
      "rules": Object {},
    },
  ],
}
`;

exports[`lib not nested should provide the controller and service 1`] = `
"import { Module } from '@nestjs/common';
import { MyLibController } from './my-lib.controller';
import { MyLibService } from './my-lib.service';

@Module({
	controllers: [MyLibController],
	providers: [MyLibService],
	exports: [MyLibService],
})
export class MyLibModule {}
"
`;

exports[`lib not nested should provide the controller and service 2`] = `
"import { Controller } from '@nestjs/common';
import { MyLibService } from './my-lib.service';

@Controller('my-lib')
export class MyLibController {
  constructor(private myLibService: MyLibService) {}
}
"
`;

exports[`lib not nested should provide the controller and service 3`] = `
"export * from './lib/my-lib.controller';
export * from './lib/my-lib.service';
export * from './lib/my-lib.module';

"
`;
